ahnlich/ai/src/engine/ai/providers/ort.rs

```diff
@@ -0,0 +1,266 @@
+                repo_name: "Qdrant/resnet50-onnx".to_string(),
+                weights_file: "model.onnx".to_string(),
+                input_param: "input".to_string(),
+                output_param: "image_embeds".to_string(),
+                ..Default::default()
+            }),
+            SupportedModels::ClipVitB32 => Ok(ORTImageModel {
+                repo_name: "Qdrant/clip-ViT-B-32-vision".to_string(),
+                weights_file: "model.onnx".to_string(),
+                input_param: "pixel_values".to_string(),
+                output_param: "image_embeds".to_string(),
+                ..Default::default()
+            }),
+            _ => Err(AIProxyError::AIModelNotSupported),
+        };
+        Ok(ORTModel::Image(model_type?))
+    }
+}
+
+impl ORTProvider {
+    pub(crate) fn new() -> Self {
+        Self {
+            cache_location: None,
+            cache_location_extension: PathBuf::from("huggingface"),
+            supported_models: None,
+            model: None,
+        }
+    }
+
+    pub fn normalize(v: &[f32]) -> Vec<f32> {
+        let norm = (v.iter().map(|val| val * val).sum::<f32>()).sqrt();
+        let epsilon = 1e-12;
+
+        // We add the super-small epsilon to avoid dividing by zero
+        v.iter().map(|&val| val / (norm + epsilon)).collect()
+    }
+}
+
+impl ProviderTrait for ORTProvider {
+    fn set_cache_location(&mut self, location: &Path) -> &mut Self {
+        self.cache_location = Some(location.join(self.cache_location_extension.clone()));
+        self
+    }
+
+    fn set_model(&mut self, model: &SupportedModels) -> &mut Self {
+        self.supported_models = Some(*model);
+        self
+    }
+
+    fn load_model(&mut self) -> &mut Self {
+        let cache_location = self
+            .cache_location
+            .clone()
+            .expect("Cache location not set.");
+        let supported_model = self.supported_models.expect(
+            &AIProxyError::AIModelNotInitialized.to_string());
+        let ort_model = ORTModel::try_from(&supported_model).unwrap();
+
+        let threads = available_parallelism()
+            .expect("Could not find the threads")
+            .get();
+
+        let cache = Cache::new(cache_location);
+        let api = ApiBuilder::from_cache(cache)
+            .with_progress(true)
+            .build()
+            .unwrap();
+
+        match ort_model {
+            ORTModel::Image(ORTImageModel {
+                weights_file,
+                repo_name,
+                input_param,
+                output_param,
+                ..
+            }) => {
+                let model_repo = api.model(repo_name.clone());
+                let model_file_reference = model_repo
+                    .get(&weights_file)
+                    .unwrap_or_else(|_| panic!("Failed to retrieve {} ", weights_file));
+                let executioners: Vec<ExecutionProviderDispatch> = Default::default();
+                let session = Session::builder()
```

Comment:
We can make Session builder errors its own variant which we move upwards with `?`